jupyter_collaboration/app.py

```diff
@@ -112,6 +122,39 @@ def initialize_handlers(self):
             ]
         )
 
+    async def get_document(
+        self: YDocExtension,
+        *,
+        path: str,
+        content_type: Literal["notebook", "file"],
+        file_format: Literal["json", "text"],
+    ) -> YBaseDoc | None:
+        """Get a read-only view of the shared model for the matching document.
+
+        The returned shared model is a fork, meaning that any changes made to it
+        will not be propagated to the shared model used by the application.
+        """
+        file_id_manager = self.serverapp.web_app.settings["file_id_manager"]
+        file_id = file_id_manager.index(path)
+
+        encoded_path = encode_file_path(file_format, content_type, file_id)
+        room_id = room_id_from_encoded_path(encoded_path)
+
+        try:
+            room = await self.ywebsocket_server.get_room(room_id)
+        except RoomNotFound:
+            return None
+
+        if isinstance(room, DocumentRoom):
+            update = room.ydoc.get_update()
+
+            fork_ydoc = Doc()
+            fork_ydoc.apply_update(update)
+
+            return YDOCS.get(content_type, YDOCS["file"])(fork_ydoc)
```

Comment:
So this will be a snapshot of the document at the time this method is called, but don't you want a live document, with all future updates applied?

Reply:
Good point. For my use case I am fine with having just a snapshot, but maybe having a proper _view_ which gets the future updates would make more sense for others?

Reply:
> I don't know if you would be fine with having a live fork.

Maybe.

I do not _strictly_ want a snapshot. I am indifferent to that, as long as it does not force me to wait for the lock/update/etc.

> In the latter case, I'm wondering why you still need a shared document, rather than just converting it to a plain Python object?

Converting a notebook to a plain Python object takes 4 times longer (https://github.com/jupyterlab/jupyter-collaboration/issues/270#issuecomment-2027549354). 30 ms on a notebook with 100 empty cells makes me worried that this would be unsuitable for the completions use case especially when notebooks have large visualisations in outputs etc. Often we will only want last/next ~10 cells so we would not be converting full notebook to Python ever.

Reply:
I do not follow. I am not in the frontend, I am in the server extension. (10 cells is an underestimate/figure of speech, it will be probably much more, the point is it will not be all 5000 cells in a notebook so long but a fraction, maybe 50 maybe 200)

Reply:
Ah ok, then you could do the same in the backend: get the shared document and convert any subset of it you'd like.

Reply:
This is exactly what I am doing in https://github.com/jupyterlab/jupyter-ai/pull/708 and plan to use this PR for. The question is do we want to return:
- a frozen read-only model copy
- a live read-only model
- a live read-write model

Again, for my use case this is really indifferent. From the point of view of API design I thought that starting with the smallest possible API surface (frozen model copy) and then expanding it in backward-compatible way (adding live updates, adding write access) seems like an easy decision (well maybe that is not fully backward compatible, but it would not be breaking for most use cases).

Currently this PR returns a frozen (but not read-only) copy of the model. I believe I can make it read-only by monkey-patching some methods in the Model instance after creating the copy so that they raise an exception when write is attempted.

What is the best next step/solution in your opinion?

Reply:
> but I don't understand the benefit of having a frozen copy of the model

There is not any - other than that it is a simple way of ensuring that other extensions do not start writing to the shared model via this new public API.

> I think `get_document()` should just return the shared model, not a copy of it.

I am happy with that.

Reply:
What do you think about adding a parameter `get_document(copy=True)`? In the future we could also add `live=True`, which would correspond to the use-case we have in the suggestion system.

Reply:
Sounds fine to me, done in: https://github.com/jupyterlab/jupyter-collaboration/pull/275/commits/3a8ba7ddcbd179b65618bd574c306f346ead57fa